### PR TITLE
[21363] Migration guide

### PR DESCRIPTION
**Pull request by elianalf**
## Description
This PR updates the Upgrading.md document, which aims to help during the migration process from eProsima *Fast DDS version* 2 to *Fast DDS version* 3.
It contains the changes that a user may need to apply to migrate to *Fast DDS v3.0.0*.

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- N/A Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- N/A Any new/modified methods have been properly documented using Doxygen. 
- N/A Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- N/A Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- N/A Changes are API compatible. 
- N/A New feature has been added to the `versions.md` file (if applicable).
- N/A New feature has been documented/Current behavior is correctly described in the documentation. 
    
- N/A Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- _N/A_; If this is a critical bug fix, backports to the critical-only supported branches have been requested.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: UPGRADING.md [hunk 1]
```diff
@@ -1,6 +1,294 @@
...
+* `heartbeatResponseDelay` is `heartbeat_response_delay`.
+* `initialAcknackDelay` is `initial_acknack_delay`.
+* `expectsInlineQos` is `expects_inline_qos`.
+* `m_typeSize` is `max_serialized_type_size`
+* `m_isGetKeyDefined` is `is_compute_key_provided`
+* `m_topicDataTypeName` is `topic_data_typename_`
+
+
+## Examples
+
```
Comment (JesusPoderoso) [inline UPGRADING.md:228]:
```suggestion
## Examples

All the examples have been refactored to follow the same structure:
* File names, guards, and classes follow new format.
* Detailed and well-formed README.md with example explanation.
* Example structured in applications, stopped by `SIGTERM` signal.
```

File: UPGRADING.md [hunk 2]
```diff
@@ -1,6 +1,294 @@
...
+* All `eprosima::fastrtps::` namespace references were moved to `eprosima::fastdds::`.
+* `SubscriptionBuiltinTopicData`, `PublicationBuiltinTopicData` and `ParticipantBuiltinTopicData` were moved from `fastdds::dds::builtin::` to `fastdds::dds::`.
+* `EventKindBits::` references changed to`EventKind::`.
+* `EventKindEntityId::` references changed to`EntityId::`.
+* `StatisticsEventKind::` references changed to `statistics::EventKind::`.
+* `Duration_t` and `c_TimeInfinite` reference were moved to `dds::`.
+* `ParticipantAttributes`, `PublisherAttributes`, `SubscriberAttributes`, `ReplierAttributes`, `RequesterAttributes` were moved to `xmlparser::`.
+* `Time_hpp` references were divided into `dds::Time_t::` and `rtps::Time_t::`.
+
+
```
Comment (JesusPoderoso) [inline UPGRADING.md:55]:
```suggestion
```

File: UPGRADING.md [hunk 3]
```diff
@@ -1,6 +1,294 @@
...
+| fastrtps/qos/SampleRejectedStatus.hpp | fastdds/dds/core/status/SampleRejectedStatus.hpp |
+| fastrtps/participant/Participant.h | fastdds/rtps/participant/RTPSParticipant.hpp |
+| fastrtps/transport/TCPv4TransportDescriptor.h | fastdds/rtps/transport/TCPv4TransportDescriptor.hpp |
+| fastrtps/transport/TCPv6TransportDescriptor.h| 	fastdds/rtps/transport/ TCPv6TransportDescriptor.hpp |
+| fastrtps/transport/UDPv4TransportDescriptor.h | fastdds/rtps/transport/ UDPv4TransportDescriptor.hpp |
+| fastrtps/transport/UDPv6TransportDescriptor.h | fastdds/rtps/transport/ UDPv6TransportDescriptor.hpp |
+| fastrtps/transport/UDPTransportDescritpor.h | fastdds/rtps/transport/UDPTransportDescritpor.hpp |
+| fastrtps/transport/TCPTransportDescritpor.h | fastdds/rtps/transport/TCPTransportDescritpor.hpp |
+
+
```
Comment (JesusPoderoso) [inline UPGRADING.md:85]:
```suggestion
```

File: UPGRADING.md [hunk 4]
```diff
@@ -1,6 +1,294 @@
...
+| char* TopicDataType::getName() | string& TopicDataType::get_name() |
+| TypeSupport::calculate_serialized_size_provider(const void* const, `DataRepresentationId_t`) | TypeSupport::calculate_serialized_size(const void* const, `DataRepresentationId_t`) |
+|   get_key(void, `InstanceHandle_t`*, bool) | compute_key(`SerializedPayload_t`&, `InstanceHandle_t`&, bool) |
+| DynamicPubSubType::createData() | DynamicPubSubType::create_data() |
+| DynamicPubSubType::deleteData(void*) | DynamicPubSubType::delete_data(void*) |
+| DynamicPubSubType::getKey(const void* const, `InstanceHand`*, bool) | DynamicPubSubType::compute_key(const void* const, `InstanceHand`&, bool) |
+| DynamicPubSubType::getSerializedSizeProvider(const void* const, `DataRepresentationId_t`) | DynamicPubSubType::calculate_serialized_size(const void* const, `DataRepresentationId_t`) |
+
+
+
```
Comment (JesusPoderoso) [inline UPGRADING.md:206]:
```suggestion
```

File: UPGRADING.md [hunk 5]
```diff
@@ -1,6 +1,294 @@
...
+### Hello World
+Refactor the HelloWorld example with the current new example format.
+In this hello world example, the key changes are:
+* Removed --env option and set that environment behavior as default.
+* Added subscriber waitsets class and its usage.
+* Provided 	XML profiles examples that target several scenarios (e.g.,SampleConfig_Controller, Events, Multimedia).
+
+### X-Types Examples
+In this X-Types example, a type is defined at runtime on the publisher side using the Dynamic Types API, and the subscriber discovers the type, creates a reader for it, and prints the received data. This example is type compatible with the Hello World example, and a compatibility test has been added for this.
```
Comment (JesusPoderoso) [inline UPGRADING.md:237]:
```suggestion
In this X-Types example, a type is defined at runtime on the publisher side using the Dynamic Types API, and the subscriber discovers the type, creates a reader for it, and prints the received data. 
This example is type compatible with the Hello World example, and a compatibility test has been added for this.
```

File: UPGRADING.md [hunk 6]
```diff
@@ -1,6 +1,294 @@
...
+### Configuration
+In this configuration example, the key changes are:
+* Included LargeData as an option (builtin transport argument).
+* Included all previous QoS examples:
+    * Deadline
+    * Disable positive ACKs
+    * Lifespan
+    * Liveliness
+    * Ownership (strength)
+    * Partitions
```
Comment (JesusPoderoso) [inline UPGRADING.md:248]: Try to write this as a parragraph, as the suggestion in helloworld example. This applies also to the remain examples

File: UPGRADING.md [hunk 7]
```diff
@@ -1,6 +1,294 @@
...
+
+### Delivery Mechanisms
+In this delivery mechanisms example, the key changes are:
+* Loans and data-sharing compatible: bounded types, final extensibility.
+* Loans mechanism for data management.
+* Option to select all delivery mechanisms.
+
+### Discovery Server
+Refactor the DiscoveryServerExample example with the current new example format.
+This refactors the DiscoveryServerExample to the new example format. It also adds proper tests for the examples testing automation.
```
Comment (JesusPoderoso) [inline UPGRADING.md:269]:
```suggestion
This refactors the DiscoveryServerExample to the new example format. 
It also adds proper tests for the examples testing automation.
```

File: UPGRADING.md [hunk 8]
```diff
@@ -1,6 +1,294 @@
...
+|   get_key(void, `InstanceHandle_t`*, bool) | compute_key(`SerializedPayload_t`&, `InstanceHandle_t`&, bool) |
+| DynamicPubSubType::createData() | DynamicPubSubType::create_data() |
+| DynamicPubSubType::deleteData(void*) | DynamicPubSubType::delete_data(void*) |
+| DynamicPubSubType::getKey(const void* const, `InstanceHand`*, bool) | DynamicPubSubType::compute_key(const void* const, `InstanceHand`&, bool) |
+| DynamicPubSubType::getSerializedSizeProvider(const void* const, `DataRepresentationId_t`) | DynamicPubSubType::calculate_serialized_size(const void* const, `DataRepresentationId_t`) |
+
+
+
+## Struct, Enum, Variable
+
```
Comment (JesusPoderoso) [inline UPGRADING.md:208]: Add some link sentence or small paragraph here

**Review by JesusPoderoso — CHANGES_REQUESTED** [review]
Partial review (style mainly).
Pending deep review about the exposed content (check that there is nothing missing to be exposed in this guide)

File: UPGRADING.md (hunk 6 above)
Comment (elianalf) [reply UPGRADING.md:248]: Don't you think it's clearer as a bullet list?

File: UPGRADING.md [hunk 9]
```diff
@@ -1,6 +1,300 @@
...
+    * fastdds.manifest.in
+    * fastdds-config.cmake
+
+* The fastrtps.rc file has been renamed as fastdds.rc.
+* Deprecated APIs marked with `FASTDDS_DEPRECATED` and `FASTDDS_TODO_BEFORE` macros have been removed.
+
+## Compatibility with Fast CDR
+
+Fast DDS v3 is only compatible with Fast CDR v2.
+If you are not using Fast CDR as [third-party](https://fast-dds.docs.eprosima.com/en/latest/installation/configuration/cmake_options.html#third-party-libraries-options), please ensure updating your local dependencies.
```
Comment (EduPonz) [inline UPGRADING.md:40]:
```suggestion
If you are not using Fast CDR as [third-party](https://fast-dds.docs.eprosima.com/en/latest/installation/configuration/cmake_options.html#third-party-libraries-options), please ensure that your local dependencies are up-to-date.
```

File: UPGRADING.md [hunk 10]
```diff
@@ -1,6 +1,300 @@
...
+The following list contains the namespace changes and migrations:
+
+* All `eprosima::fastrtps::` namespace references were moved to `eprosima::fastdds::`.
+* `SubscriptionBuiltinTopicData`, `PublicationBuiltinTopicData` and `ParticipantBuiltinTopicData` were moved from `fastdds::dds::builtin::` to `fastdds::dds::`.
+* `EventKindBits::` references changed to`EventKind::`.
+* `EventKindEntityId::` references changed to`EntityId::`.
+* `StatisticsEventKind::` references changed to `statistics::EventKind::`.
+* `Duration_t` and `c_TimeInfinite` reference were moved to `dds::`.
+* `ParticipantAttributes`, `PublisherAttributes`, `SubscriberAttributes`, `ReplierAttributes`, `RequesterAttributes` were moved to `xmlparser::`.
+* `Time_t.hpp` references were explicitly distingueshed between `dds::Time_t::` and `rtps::Time_t::`.
```
Comment (EduPonz) [inline UPGRADING.md:53]: We already had two classes, I think what we need to say is that they are in different files now

File: UPGRADING.md [hunk 11]
```diff
@@ -1,6 +1,300 @@
...
+* ChangeForReader.hpp
+* ReaderLocator.hpp
+* ReaderProxy.hpp
+* ServerAttributes.hpp
+* TopicAttributes.hpp
+* TypeLookupService.hpp
+
+## API changes
+
+The table contains the list of API changes, showing the previous methods and the corresponding new ones introduced in Fast DDS v3.
```
Comment (EduPonz) [inline UPGRADING.md:137]:
```suggestion
The table below contains the list of API changes, showing the previous methods and the corresponding new ones introduced in Fast DDS v3.
```

File: UPGRADING.md [hunk 12]
```diff
@@ -1,6 +1,300 @@
-Upgrading from previous versions
-================================
+# Migration Guide from Fast DDS v2 to Fast DDS v3
 
-If you have used *fastddsgen* to generate part of your code, it is always *recommended* to regenerate it.
+This document aims to help during the migration process from eProsima *Fast DDS version* 2 to *Fast DDS version* 3.
+For more information about all the updates, please refer to the [release notes](https://fast-dds.docs.eprosima.com/en/latest/notes/notes.html).
 
-When upgrading from a version older than 1.7.0, regeneration is **required**.
+As announced in previous minor releases, it is always *recommended* to regenerate the types with the latest *eProsima Fast DDS Gen* release.
```
Comment (EduPonz) [inline UPGRADING.md:6]: It's not always the latest! It's the latest patch associated with the corresponding Fast DDS version

File: UPGRADING.md [hunk 13]
```diff
@@ -1,6 +1,300 @@
...
+- [Public headers migrated to *fastdds*](#public-headers-migrated-to-fastdds)
+- [Public headers moved to private](#public-headers-moved-to-private)
+- [API changes](#api-changes)
+- [Struct, Enum, Variable](#struct-enum-variable)
+- [Examples](#examples)
+
+
+## Library management
+
+The below list expose the changes related to the package name, environment variables and other library usages.
```
Comment (EduPonz) [inline UPGRADING.md:23]:
```suggestion
The list below exposes the changes related to the package name, environment variables and other library usages.
```

File: UPGRADING.md [hunk 14]
```diff
@@ -1,6 +1,300 @@
...
+
+* The package has been renamed from `fastrtps` to `fastdds`.
```
Comment (EduPonz) [inline UPGRADING.md:25]:
```suggestion
* The CMake project has been renamed from `fastrtps` to `fastdds`.
```

File: UPGRADING.md [hunk 15]
```diff
@@ -1,6 +1,300 @@
...
+* XML profiles loading environment variable has been renamed to: `FASTDDS_DEFAULT_PROFILES_FILE`.
+* The configuration file that Fast DDS looks for to load the profiles has been renamed to `DEFAULT_FASTDDS_PROFILES.xml`.
+* XML Schema namespace in fastdds_profiles.xsd has been updated to http://www.eprosima.com.
```
Comment (EduPonz) [inline UPGRADING.md:28]:
```suggestion
* XML Schema namespace in `fastdds_profiles.xsd` has been updated to http://www.eprosima.com.
```

File: UPGRADING.md [hunk 16]
```diff
@@ -1,6 +1,300 @@
...
+## Namespace migrations and changes
+
+The following list contains the namespace changes and migrations:
+
+* All `eprosima::fastrtps::` namespace references were moved to `eprosima::fastdds::`.
+* `SubscriptionBuiltinTopicData`, `PublicationBuiltinTopicData` and `ParticipantBuiltinTopicData` were moved from `fastdds::dds::builtin::` to `fastdds::dds::`.
+* `EventKindBits::` references changed to`EventKind::`.
+* `EventKindEntityId::` references changed to`EntityId::`.
+* `StatisticsEventKind::` references changed to `statistics::EventKind::`.
+* `Duration_t` and `c_TimeInfinite` reference were moved to `dds::`.
```
Comment (EduPonz) [inline UPGRADING.md:51]:
```suggestion
* `Duration_t` and `c_TimeInfinite` references were moved to `dds::`.
```

File: UPGRADING.md [hunk 17]
```diff
@@ -1,6 +1,300 @@
...
+* `ParticipantAttributes`, `PublisherAttributes`, `SubscriberAttributes`, `ReplierAttributes`, `RequesterAttributes` were moved to `xmlparser::`.
```
Comment (EduPonz) [inline UPGRADING.md:52]: They are in fact private now. This list is missing `TopicAttributes`

File: UPGRADING.md [hunk 18]
```diff
@@ -1,6 +1,300 @@
...
+* `Time_t.hpp` references were explicitly distingueshed between `dds::Time_t::` and `rtps::Time_t::`.
+
+## Public headers migrated to *fastdds*
+
+All public header extensions have been changed to `.hpp`.
```
Comment (EduPonz) [inline UPGRADING.md:57]: As this is the first time we say "public header", please clarify that those are the ones under `include`

**Review by EduPonz — CHANGES_REQUESTED** [review]

File: UPGRADING.md (hunk 17 above)
Comment (elianalf) [reply UPGRADING.md:52]: Ok so I will remove them from this list and I will keep only the public ones

File: UPGRADING.md [hunk 19]
```diff
@@ -1,6 +1,299 @@
-Upgrading from previous versions
-================================
+# Migration Guide from Fast DDS v2 to Fast DDS v3
 
-If you have used *fastddsgen* to generate part of your code, it is always *recommended* to regenerate it.
+This document aims to help during the migration process from eProsima *Fast DDS version* 2 to *Fast DDS version* 3.
+For more information about all the updates, please refer to the [release notes](https://fast-dds.docs.eprosima.com/en/latest/notes/notes.html).
 
-When upgrading from a version older than 1.7.0, regeneration is **required**.
+It is always *required* to regenerate the types with the latest version of *eProsima Fast DDS Gen*, *Fast DDS Gen* v4.0.0.
```
Comment (EduPonz) [inline UPGRADING.md:6]:
```suggestion
It is always *required* to regenerate the types with the appropriate version of *eProsima Fast DDS Gen* (see [products compatibility](https://fast-dds.docs.eprosima.com/en/latest/notes/versions.html#eprosima-products-compatibility))
```

File: UPGRADING.md [hunk 20]
```diff
@@ -1,6 +1,299 @@
...
+- [Public headers migrated to *fastdds*](#public-headers-migrated-to-fastdds)
+- [Public headers moved to private](#public-headers-moved-to-private)
+- [API changes](#api-changes)
+- [Struct, Enum, Variable](#struct-enum-variable)
+- [Examples](#examples)
+
+
+## Library management
+
+The below exposes exposes the changes related to the package name, environment variables and other library usages.
```
Comment (EduPonz) [inline UPGRADING.md:22]:
```suggestion
The list below exposes the changes related to the package name, environment variables and other library usages.
```

File: UPGRADING.md [hunk 21]
```diff
@@ -1,6 +1,299 @@
...
+
+- [Library management](#library-management)
+- [Compatibility with Fast CDR](#compatibility-with-fast-cdr)
+- [Namespace migrations and changes](#namespace-migrations-and-changes)
+- [Public headers migrated to *fastdds*](#public-headers-migrated-to-fastdds)
+- [Public headers moved to private](#public-headers-moved-to-private)
+- [API changes](#api-changes)
+- [Struct, Enum, Variable](#struct-enum-variable)
+- [Examples](#examples)
+
```
Comment (EduPonz) [inline UPGRADING.md:18]:
```suggestion
```

File: UPGRADING.md [hunk 22]
```diff
@@ -1,6 +1,299 @@
...
+* `SubscriptionBuiltinTopicData`, `PublicationBuiltinTopicData` and `ParticipantBuiltinTopicData` were moved from `fastdds::dds::builtin::` to `fastdds::dds::`.
+* `EventKindBits::` references changed to`EventKind::`.
+* `EventKindEntityId::` references changed to`EntityId::`.
+* `StatisticsEventKind::` references changed to `statistics::EventKind::`.
+* `Duration_t` and `c_TimeInfinite` references were moved to `dds::`.
+* `Time_t.hpp` references were moved from `eprosima::fastdds::` to `eprosima::fastdds::dds`.
+
+## Public headers migrated to *fastdds*
+
+All the extensions of the headers in /include (public headers) have been changed to `.hpp`.
```
Comment (EduPonz) [inline UPGRADING.md:55]:
```suggestion
All the extensions of the headers under `include`, that is public headers that applications can include, have been changed to `.hpp`.
```

**Review by EduPonz — CHANGES_REQUESTED** [review]

**Review by EduPonz — APPROVED** [review]

**Pull request merged**